job_creator.php

```diff
@@ -0,0 +1,239 @@
+<?php
+
+class JobCreator
+{
+    public function getInstallerVersion($githubRepository, $branch)
+    {
+        $repo = explode('/', $githubRepository)[1];
+        if (in_array($repo, NO_INSTALLER_REPOS)) {
+            return '';
+        }
+        $v = explode('.', $branch);
+        if (count($v) == 1) {
+            return '4.x-dev';
+        }
+        if (in_array($repo, LOCKSTEPED_REPOS)) {
+            return '4' . $v[1] . 'x-dev';
+        } else {
+            // use the latest minor version of installer
```

Comment:
We have `REQUIRE_RECIPE` in travis now for this purpose, and it gets out of date very quickly and causes issues.  One of my goals with this is to put more smarts into the dynamic matrix generation so that we always get the correct version of installer

That example is nonsensical. If removed lockstepping altogether then the correct solution is to update gha-generate-matrix, not to hardcode things into module workflow yml files

Reply:
> it gets out of date very quickly

It gets out of date because it was put in there as a stop-gap with the assumption that when we move to github actions it will be unnecessary, because we would have a fully dynamic approach. But I don't think we will ever have a fully dynamic approach that works in every situation.

We already update composer.json every release and I wouldn't be surprised if we automate that at some stage. I don't see this as being any more difficult to update than that.

We _especially_ won't be able to cover community modules dynamically, so even if that option is _only added for the community_ IMO we should add it. Like I said before, that may be a stretch goal but let's stretch for it where we can.

> If removed lockstepping altogether then the correct solution is to update gha-generate-matrix

So we'd have a map of module versions to installer versions? Something like this?
```php
const MODULE_TO_INSTALLER = [
    'silverstripe-config' => [
        '1.4' => '4.11',
    ],
];
```

**In any case** ultimately as I have mentioned I think this option is necessary to allow community modules to use this CI and to fulfil the AC that says "There's a way to override the defaults" - _even if we don't use it in our own modules_.

Reply:
No, I think this only promotes bad habits and tech debt.  If allow defining a specific installer for community modules, then this what'll end up happening:

```
<regular yml config to include ci>
  with:
    installer_version: ^4.10
```

And it'll work for the time being, and then it'll stay like that pretty much forever.  So in 2 years time when this version of installer is EOL, it'll still be what's used to test the community module, instead of what it should be testing against which is the latest version of installer.  It's far better to get the latest version of installer and have it go red so people know the module is actually broken.

>> So we'd have a map of module versions to installer versions?

Even if we were to remove lockstepping, we're not making that decisions as part of this PR, so we don't need to work out the implementation details right now

Reply:
Fair enough. We'll see what issues we run into and address them as they come. Good point about community modules that wouldn't update their installer version config.